checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -1,176 +1,870 @@
+          continue;
+        }
+        computeOwningForParamOfCurrentMethod(obligations, paramsOfCurrentMethod, invocation, arg);
+      }
+    }
+  }
 
-    } else {
-      if (!(cur instanceof SingleSuccessorBlock)) {
-        throw new BugInCF("BlockImpl is neither a conditional block nor a SingleSuccessorBlock");
+  /**
+   * Computes an @Owning annotation for a field or parameter passed in the varargs argument
+   * position.
+   *
+   * @param obligations set of obligations associated with the current block
+   * @param paramsOfCurrentMethod the parameters of the current method
+   * @param invocation the method invocation node to check
+   * @param varArgsNode the VarArg node of the given method invocation node
+   * @param arg the argument of a method invocation node
+   */
+  private void computeOwningParamsForVarArgs(
+      Set<Obligation> obligations,
+      List<? extends VariableTree> paramsOfCurrentMethod,
+      MethodInvocationNode invocation,
+      ArrayCreationNode varArgsNode,
+      Node arg) {
+    for (Node varArgNode : varArgsNode.getInitializers()) {
+      Element varArgElt = TreeUtils.elementFromTree(varArgNode.getTree());
+
+      if (varArgElt == null) {
+        continue;
       }
 
-      Block b = ((SingleSuccessorBlock) cur).getSuccessor();
-      if (b != null) {
-        successorBlock.add(b);
+      if (varArgElt.getKind().isField()) {
+        inferOwningField(varArgNode, invocation);
+      } else {
+        computeOwningForParamOfCurrentMethod(obligations, paramsOfCurrentMethod, invocation, arg);
       }
     }
-    return successorBlock;
+  }
+
+  /**
+   * Computes an @Owning annotation for any parameter of the current method that is aliased with the
+   * argument passed to the method call.
+   *
+   * @param obligations set of obligations associated with the current block
+   * @param paramsOfCurrentMethod the parameters of the current method
+   * @param invocation the method invocation node to check
+   * @param arg the argument of a method invocation node
+   */
+  private void computeOwningForParamOfCurrentMethod(
+      Set<Obligation> obligations,
+      List<? extends VariableTree> paramsOfCurrentMethod,
+      MethodInvocationNode invocation,
+      Node arg) {
+
+    for (int i = 0; i < paramsOfCurrentMethod.size(); i++) {
+      if (resourceLeakAtf.hasEmptyMustCallValue(paramsOfCurrentMethod.get(i))) {
+        continue;
+      }
+      VariableTree currentMethodParamTree = paramsOfCurrentMethod.get(i);
+      VariableElement currentMethodParamElt =
+          TreeUtils.elementFromDeclaration(currentMethodParamTree);
+      if (isNodeAndElementAliased(obligations, arg, currentMethodParamElt)) {
+        JavaExpression paramJe = JavaExpression.fromVariableTree(currentMethodParamTree);
+        if (mustCallObligationSatisfied(invocation, currentMethodParamElt, paramJe)) {
+          addOwningToParam(i + 1);
+          break;
+        }
+      }
+    }
+  }
```

Comment:
I feel there is significant duplication between this code and the code at lines 502--520.  Can we restructure the logic to get rid of this duplication, maybe by creating a common method?  Please look for other similar opportunities where there is this much code that is so similar